[Canary] Transcoding Support + More Image Format Support (JPEGXL)

maxpiva opened the pull request:
Canary testers, I need your help! This is a massive PR that swaps out the Image system in Kavita and brings Transcoding support. This is built from the nightly. Please give it a serious stress test and report back on the [nightly thread](https://discord.com/channels/821879810934439936/1316389775426588732).

# Added
 - Added: Added support for JPEG-XL, JPEG 200, and HEIF image formats
 - Added: Kavita now has transcoding support. This means Kavita will respect Accept Headers and if the target browser/device doesn't support the image, Kavita will convert it on the fly. From testing this is not noticeable and without quality drop.
 
 # Changed
- Changed: Kavita should no longer require SSE4.2 CPU Instructions and should be able to run on older hardware. 
- Changed: Due to the massive overhaul to the image manipulation library, smart cropping (mainly visible on oddly-shaped cover images) was changed from NetVips to SmartCrop.js C# library. Minor differences from testing. 

# Developer
- Swapped out NetVips with ImageMagick (Magick.NET)
- Updated Flurl.Http with a new configuration 
- Removed NetVips and SixLabors.ImageSharp

maxpiva committed:
Add JXL Support.

maxpiva committed:
Make sure if djxl do not exists, continue as usual.
Fixed test and benchmark DI.
Point docker-build to another hub
Edit dockerfile and include jxl package.

maxpiva committed:
Changed the way conversion happens.
1. Supported Image Formats came in the accept flag from the browser.
2. If jxl (or future formats, ie browser not supported AVIF, HEIF, etc) are not supported, Kavita will convert it to a suitable display format via conversion providers.
3. if it supported it will return the image as is, no conversion needed.
4. Thumbnails are always converted.

maxpiva committed:
Changed the way conversion happens.
1. Supported Image Formats came in the accept flag from the browser.
2. If jxl (or future formats, ie browser not supported AVIF, HEIF, etc) are not supported, Kavita will convert it to a suitable display format via conversion providers.
3. if it supported it will return the image as is, no conversion needed.
4. Thumbnails are always converted.

maxpiva committed:
Missing file

maxpiva committed:
Merge branch 'Kareadita:develop' into develop

maxpiva committed:
Revert "Changed the way conversion happens."

This reverts commit 7f6db2439cfe00e6051ea3237b621ed0ec64054f.

maxpiva committed:
Reapply "Changed the way conversion happens."

This reverts commit 9d48664b08f01303f1527bc1c577e1d4376f89ff.

maxpiva committed:
Merge with Karedita

maxpiva committed:
Add Documentation

maxpiva committed:
FIX BOMS

maxpiva committed:
Recode SupportedImageTypesFromRequest removing possible errors.

maxpiva committed:
Remove own docker build

maxpiva committed:
Shell Call no longer used, since swapped image conversion to Image Magick. Rollback IOsInfo.cs

maxpiva committed:
Merge branch 'Kareadita:develop' into JPEG-XL_JPG2000_HEIF

maxpiva committed:
Edit CodeStyle
Fix GetCoverImage, trying to create a temporary image in current directory, and rename the parameter field name, that generates a wrong asumption.

maxpiva committed:
Remove custom docker builder

maxpiva committed:
Fixed discovered issues.
Optimize Cover Creation

maxpiva committed:
Merge branch 'Kareadita:develop' into JPEG-XL_JPG2000_HEIF

maxpiva committed:
Merge branch 'Kareadita:develop' into JPEG-XL_JPG2000_HEIF

maxpiva committed:
Removed NetVips and ImageSharp (Use only ImageMagick)
Removing ImageConvertors (No Longer Needed)
Refactored mime types and extensions constants

maxpiva committed:
Remove CropFromDimensions (Geometry will take care)

maxpiva committed:
BOM Changes

maxpiva committed:
Delete .bak

maxpiva committed:
Bug fixes.

maxpiva committed:
Merge branch 'Kareadita:develop' into RemoveSixLaborsAndNetVips

maxpiva committed:
Added SmartCrop (port from smartcrop.js)

maxpiva committed:
Refactor and cleanup.

maxpiva committed:
BOM Fixes

maxpiva committed:
Update docker-build.sh

Enable back docker build/push

adamradocz commented:
HEIF and AVIF are superior formats compared to JpegXL and Jpeg2000. Why the hassle to support them?

maxpiva commented:
> HEIF and AVIF are superior formats compared to JpegXL and Jpeg2000. Why the hassle to support them?

Apple might want to talk to you, while I don't want to start a format war, I want to know how much the google ~~chrome~~ AVIF team is paying you, I want the same deal! :)

Support came from changing netvips to ImageMagick we don't need to add specific support/mantain for JPEG-XL. There is a secondary why, @majora2007 probably can explain it better, but netvips limits the possible CPU targets.

What is added, is, we identify if your browser supports a new format A, B, C or D, and if not with transcode the image to a format that your browser/app supports. 

While is true that the original intent was adding JPEG-XL, let's say because my own bias, and current is in vogue. 

In the end, we end supporting AVIF, HEIF, JPEG-XL and JPEG2000 comics/manga, regardless your browser/app supports it or not.

majora2007 commented on API.Benchmark/ArchiveServiceBenchmark.cs:
```diff
@@ -24,16 +24,16 @@ public class ArchiveServiceBenchmark
-    private readonly PngEncoder _pngEncoder = new PngEncoder();
-    private readonly WebpEncoder _webPEncoder = new WebpEncoder();
+    private readonly IImageFactory _imageFactory;
     private const string SourceImage = "C:/Users/josep/Pictures/obey_by_grrsa-d6llkaa_colored_by_me.png";
```
We should probably remove this hardcoded path and point to an arbitrary image in TestData

maxpiva replied:
Done

maxpiva committed:
Fixed Benchmarks

maxpiva committed:
Merge branch 'RemoveSixLaborsAndNetVips' of https://github.com/maxpiva/Kavita into RemoveSixLaborsAndNetVips

maxpiva committed:
Fixed BOM

majora2007 commented on API/Entities/Enums/EncodeFormat.cs:
```diff
@@ -2,12 +2,16 @@
     [Description("AVIF")]
-    AVIF = 2
+    AVIF = 2,
+    //Internal Use
```
Let's /// this so it shows in the IDE that this is internal only.

maxpiva replied:
Done

majora2007 commented on API/Extensions/HttpExtensions.cs:
```diff
@@ -53,4 +57,51 @@ public static void AddCacheHeader(this HttpResponse response, string filename, i
+                {
+                    AddExtension(supportedExtensions, "svg");
+                }
+                else
```
Please fix the formatting in this file. 

Space after start of a comment
Newline after 84 to help break up the code
Use var whenever possible
mimeimagepart should be mimeImagePart
Newline before return on line 105

maxpiva replied:
Done

majora2007 commented on API/Extensions/HttpExtensions.cs:
```diff
@@ -53,4 +57,51 @@ public static void AddCacheHeader(this HttpResponse response, string filename, i
+                {
+                    Parser.NonUniversalSupportedMimeMappings[mimeimagepart].ForEach(x => AddExtension(supportedExtensions, x));
+                }
+                else if (mimeimagepart == "svg+xml")
```
Can we not use MimeType library?

maxpiva replied:
Not sure, this is a mime mapping against multiple extensions, added documentation to the parser stating how it works.

majora2007 commented on API/Services/BookService.cs:
```diff
@@ -1273,16 +1272,15 @@ private string GetPdfCoverImage(string fileFilePath, string fileName, string out
+        IImage image = _imageService.ImageFactory.CreateFromBGRAByteArray(rawBytes, width, height);
         stream.Seek(0, SeekOrigin.Begin);
-        image.SaveAsPng(stream);
+        image.Save(stream, EncodeFormat.PNG, 100);
```
You can leave off the 100 since it's default quality.

maxpiva replied:
New IImage interface,  do not have default value, we can add it if you want.

majora2007 replied:
I don't see ever a need to not generate at 100 quality, so we should probably just leave it off entirely. If there is ever a need, it's easy to refactor the signature.

maxpiva replied:
Well, if I remove the 100, I cannot control JPEG image quality required below, https://github.com/Kareadita/Kavita/pull/3250#discussion_r1796091070

if you want, I can add a default value to the interface.

FYI (ImageMagick)

100 PNG = Maximum Compression. loseless
0 PNG = No Compression. loseless
100 JPG = Loseless JPG
99 - 1 = Lossy JPG
AVIF and WEBP I think they're similar to JPG.

majora2007 commented on API/Services/CacheService.cs:
```diff
@@ -91,20 +94,23 @@ public IEnumerable<FileDimensionDto> GetCachedFileDimensions(string cachePath)
         }
 
         var dimensions = new List<FileDimensionDto>();
-        var originalCacheSize = Cache.MaxFiles;
```
I can't recall why I had this here, I believe it's around buffering for NetVips right?

maxpiva replied:
I have no idea, but yes, it's related to netVIPs Constants.

maxpiva replied:
Yup, removed

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -1,82 +1,164 @@
+    /// <param name="format">The encoding format to convert the image. Default is JPG</param>
+    /// <param name="quality">The quality of the image. Default is 99.</param>
+    /// <returns>The file name with extension of the replaced image file.</returns>
+    string ReplaceImageFileFormat(string filename, List<string> supportedImageFormats = null, EncodeFormat format = EncodeFormat.JPEG, int quality = 99);
```
Can we not switch to PNG by default? PNG should be better than JPEG because we already have it as a default for the whole app. I would really like to investigate this.

maxpiva replied:
Hmm. The choice in here is JPEG, is because the encoding speed. PNG tends to be slower than JPG. Since, time in this case is essential to have a good Experience, I choose JPEG as a compromise. Anyway, this is from my own experience, not specifically from ImageMagick timings. 

Let met time log, both cases and we can compare numbers.

maxpiva replied:
Some data, backing the JPEG 99 choice:

Big JXL 48 Bit RGB Image (JXL Loading time 1.7 seconds)

Whole Conversion time (Loading & Saving)
```
WEBP - 100 -  7.3 seconds
AVIF - 100 -  4.2 seconds
PNG  - 100 - 11.1 seconds (Probably deflate max)
PNG  -   0 -  2.1 seconds (Most likely, uncompressed, big file)
JPEG -  99 -  1.9 seconds (Best bang for the buck)
JPEG - 100 -  3.4 seconds

```

So, JPEG 99, takes (1.9s - 1.7s) ~200 ms to save against the other formats.


Please NOTE:
I also added, Per Image Lock on conversion, to prevent multiple threads from processing the same image at the same time.

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -1,82 +1,164 @@
     private readonly ILogger<ImageService> _logger;
     private readonly IDirectoryService _directoryService;
     private readonly IEasyCachingProviderFactory _cacheFactory;
+    private readonly IImageFactory _imageFactory;
```
Let's move all the consts together and services below.

maxpiva replied:
Done

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -209,20 +292,18 @@ private static bool IsLikelyWideImage(int width, int height)
-    public string GetCoverImage(string path, string fileName, string outputDirectory, EncodeFormat encodeFormat, CoverImageSize size)
+
+    /// <inheritdoc/>
+    public string GetCoverImage(string path, string fileName, string outputDirectory, EncodeFormat encodeFormat, CoverImageSize size, int quality = 100)
```
Is there ever a case when we wouldn't use quality of 100? Because I can't forsee one and think we should just hardcode the quality as a const within the ImageService.

maxpiva replied:
I just standardized the service calls, so is up to the consumers to change the default values, totally subjective. If you want it back, we can.

maxpiva replied:
Tell me if you want, to change the defaults, or remove from the signature.

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -265,77 +331,57 @@ public string WriteCoverThumbnail(Stream stream, string fileName, string outputD
-            thumbnail2.WriteToFile(_directoryService.FileSystem.Path.Join(outputDirectory, filename));
-
-            return filename;
+            return string.Empty; //IDK?
```
This should still return the filename

maxpiva replied:
Done

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -265,77 +331,57 @@ public string WriteCoverThumbnail(Stream stream, string fileName, string outputD
-        return Task.FromResult(outputFile);
+        using var sourceImage = _imageFactory.Create(filePath);
+        await sourceImage.SaveAsync(outputFile, encodeFormat, quality).ConfigureAwait(false);
+        return outputFile;
```
Newline before the return.

maxpiva replied:
Done

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -265,77 +331,57 @@ public string WriteCoverThumbnail(Stream stream, string fileName, string outputD
-
-            return true;
+            var result= _imageFactory.GetDimensions(filePath);
+            if (result!=null)
```
You need to fix the spacing up around the code and the return needs to be on the same line or in {}.

maxpiva replied:
Done

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -265,77 +331,57 @@ public string WriteCoverThumbnail(Stream stream, string fileName, string outputD
-            if (info == null) return false;
-
-            return true;
+            var result= _imageFactory.GetDimensions(filePath);
```
Actually just rewrite this as `return Task.FromResult(_imageFactory.GetDimensions(filePath) != null);` and keep the try/catch

maxpiva replied:
Done

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -407,24 +453,9 @@ public async Task<string> DownloadFaviconAsync(string url, EncodeFormat encodeFo
 
             // Create the destination file path
-            using var image = Image.PngloadStream(faviconStream);
+            using var image = _imageFactory.Create(faviconStream);
```
This is for me to come back to, not sure where the encode format is being used in Create.

majora2007 replied:
I looked through the PR and don't see any place where the EncodeFormat is being used to appropriately ensure the file is outputted in the correct format. Can you help me find these cases?

maxpiva replied:
Sure

maxpiva replied:
The encoding and quality, are two lines below

           await image.SaveAsync(Path.Combine(_directoryService.FaviconDirectory, filename), encodeFormat, quality);

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -480,24 +496,11 @@ public async Task<string> DownloadPublisherImageAsync(string publisherName, Enco
-        {
-            rgbPixels.Add(new Vector3(pixels[i], pixels[i + 1], pixels[i + 2]));
-        }
+        var rgbPixels = _imageFactory.GetRgbPixelsPercentage(imagePath, 10);
```
What does 10 mean? Try to avoid magic numbers.

maxpiva replied:
10% of the size. It was hardcoded on the original colorspace.

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -739,6 +688,30 @@ public static ColorScape CalculateColorScape(string sourceFile)
+    /// <inheritdoc/>
+    public string ReplaceImageFileFormat(string filename, List<string> supportedImageFormats = null, EncodeFormat format = EncodeFormat.JPEG, int quality = 99)
+    {
+        if (CheckDirectSupport(filename, supportedImageFormats))
```
Again please don't have things like:
```
if (foo)
  bar
```

maxpiva replied:
Done. I guess. Not Sure really ;)

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -739,6 +688,30 @@ public static ColorScape CalculateColorScape(string sourceFile)
+        Match m = Regex.Match(Path.GetExtension(filename), Parser.NonUniversalFileImageExtensions, RegexOptions.IgnoreCase, Parser.RegexTimeout);
+        if (!m.Success)
+            return filename;
+        var sw = Stopwatch.StartNew();
```
We can remove the timing now.

maxpiva replied:
Done

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -739,6 +688,30 @@ public static ColorScape CalculateColorScape(string sourceFile)
+
+        using var sourceImage = _imageFactory.Create(filename);
+        sourceImage.Save(destination, format, quality);
+        File.Delete(filename);
```
We probably want try/catch around Delete statements. I've seen some OSes throw permission errors.

maxpiva replied:
Done

majora2007 commented on API/Services/ImageService.cs:
```diff
@@ -739,6 +688,30 @@ public static ColorScape CalculateColorScape(string sourceFile)
+        using var sourceImage = _imageFactory.Create(filename);
+        sourceImage.Save(destination, format, quality);
+        File.Delete(filename);
+        _logger.LogDebug("Image converted from '{Extension}' to '{format.GetExtension()}' in {ElapsedMilliseconds} milliseconds", Path.GetExtension(filename), sw.ElapsedMilliseconds);
```
We can remove this logger too

maxpiva replied:
Done

majora2007 commented on API/Services/ImageServices/ImageMagick/ImageMagickImage.cs:
```diff
@@ -0,0 +1,227 @@
+    public static IImage CreateFromBase64(string base64)
+    {
+        ImageMagickImage m = new ImageMagickImage();
+        m._image = (MagickImage)MagickImage.FromBase64(base64);
```
Space after cast

majora2007 commented on API/Services/ImageServices/ImageMagick/ImageMagickImage.cs:
```diff
@@ -0,0 +1,227 @@
+    /// <returns>An instance of <see cref="ImageMagickImage"/>.</returns>
+    public static IImage CreateFromBase64(string base64)
+    {
+        ImageMagickImage m = new ImageMagickImage();
```
We should Assert or something if we base64 is invalid or null.

maxpiva replied:
Done

majora2007 commented on API/Services/ImageServices/ImageMagick/ImageMagickImage.cs:
```diff
@@ -0,0 +1,227 @@
+            floats[i + 2] = bgraByteArray[i] << 8;
+            floats[i + 3] = bgraByteArray[i + 3] << 8;
+        }
+        ImageMagickImage m = new ImageMagickImage();
```
Newline to separate code from for loop.

maxpiva replied:
Done

majora2007 commented on API/Services/ImageServices/ImageMagick/ImageMagickImage.cs:
```diff
@@ -0,0 +1,227 @@
+        }
+        ImageMagickImage m = new ImageMagickImage();
+        m._image = new MagickImage(MagickColor.FromRgba(0, 0, 0, 0), width, height);
+        using var pixels = m._image.GetPixels();
```
Newline so the using is made clear.

maxpiva replied:
Done

majora2007 commented on API/Services/ImageServices/ImageMagick/ImageMagickImage.cs:
```diff
@@ -0,0 +1,227 @@
+        m._image = new MagickImage(MagickColor.FromRgba(0, 0, 0, 0), width, height);
+        using var pixels = m._image.GetPixels();
+        pixels.SetArea(0, 0, width, height, floats);
+        return m;
```
Newline to make clear the return as well.

maxpiva replied:
Done

majora2007 commented on API/Services/ImageServices/ImageMagick/ImageMagickImage.cs:
```diff
@@ -0,0 +1,227 @@
+    /// <param name="image">The existing <see cref="MagickImage"/>.</param>
+    public ImageMagickImage(MagickImage image)
+    {
+        _image = (MagickImage)image.Clone();
```
Space after cast

maxpiva replied:
Done

majora2007 commented on API/Services/ImageServices/ImageMagick/ImageMagickImage.cs:
```diff
@@ -0,0 +1,227 @@
+    {
+        ImageMagickImage tile = overlay as ImageMagickImage;
+        if (tile == null) return;
+        _image.Composite(tile._image, x, y, CompositeOperator.Over);
```
newline after if

maxpiva replied:
Done

majora2007 commented on API/Services/ImageServices/ImageMagick/ImageMagickImage.cs:
```diff
@@ -0,0 +1,227 @@
+    /// <inheritdoc/>
+    public float[] GetRGBAImageData()
+    {
+        float[] data = null;
```
This either needs to be an empty array or needs nullable annotation.

maxpiva replied:
Done

majora2007 commented on API/Services/Tasks/Scanner/Parser/Parser.cs:
```diff
@@ -25,15 +26,29 @@ public static class Parser
     public static readonly TimeSpan RegexTimeout = TimeSpan.FromMilliseconds(500);
 
-    public const string ImageFileExtensions = @"^(\.png|\.jpeg|\.jpg|\.webp|\.gif|\.avif)"; // Don't forget to update CoverChooser
+    public static Dictionary<string, List<string>> NonUniversalSupportedMimeMappings = new Dictionary<string, List<string>>()
```
You need to leave some documentation about when to add certain formats to each section and how it works. For example UniversalFileImageExtension means all major browsers must support the format.

maxpiva replied:
Yes, to all that I didn't comment before. 

btw, what IDE are you using, can I get your styles also? Visual Studio is Rewriting the UTF-8 BOM every time it can... Also, I notice, some .cs has BOM, but they're a few that do not have.

maxpiva replied:
Edited on rider. 🤞

majora2007 reviewed (changes requested):
I went through the whole PR. Please address comments and style changes. 

I'm worried about EncodeFormat not being taken into consideration and I want to research a way to remove JPEG as the default format as we can use PNG instead.

majora2007 commented:
I saw a comment about what IDE do I use last night, but wasn't able to find it. I use Rider. We shouldn't have BOM. It's likely due to PRs submitted by the community where it might have been added.

therobbiedavis commented:
> I saw a comment about what IDE do I use last night, but wasn't able to find it. I use Rider. We shouldn't have BOM. It's likely due to PRs submitted by the community where it might have been added. 

By default VS will save with BOM. I had this issue as well and I believe you can change this default behavior to save without BOM.

maxpiva committed:
Requested changes.
Added Per Image Lock on conversion, to prevent multiple threads from processing the same image at the same time.

maxpiva committed:
Removed bak files, fixed EncodeFormat BOM

maxpiva commented:
> > I saw a comment about what IDE do I use last night, but wasn't able to find it. I use Rider. We shouldn't have BOM. It's likely due to PRs submitted by the community where it might have been added.
> 
> By default VS will save with BOM. I had this issue as well and I believe you can change this default behavior to save without BOM.

I did just that, then some other files, got BOM when they original did NOT.

Swapped to Rider, for this, Rider  does not touch the BOM (Nor add, nor remove)

majora2007 commented:
@maxpiva if you're on discord, can you reach out to me? I want to discuss about the quality a bit more and get a better understanding. 

My biggest questions are about the quality (100, 0, 99, etc) and how to know when to use what and the time delta you provided for PNG vs JPEG for transcoding, what type of image did you use and what is the quality delta between (is it noticeable to the user)?

maxpiva committed:
Removed quality parameter everywhere.
Added Quality From EncodedFormat Extension

adamradocz commented on API/Services/ImageServices/IImage.cs:
```diff
@@ -64,15 +64,15 @@ public interface IImage : IDisposable
     /// <param name="filename">The name of the file to save the image to.</param>
     /// <param name="format">The format to save the image in.</param>
     /// <param name="quality">The quality of the saved image.</param>
```
Forgot to remove the quality comments. ;)

maxpiva replied:
Doh. :O Done

majora2007 commented:
Just remembering that this impacts the docker image and that might mean we need to check with unraid/lsio/qnap where they make customizations.

@DieselTech can you confirm?

If this is true, we will need to pair this with a release that isn't primarily bugfixes (which is what the upcoming one is).

maxpiva commented:
Working in unraid as is, using (original unraid template) but changed the image path to my docker repo.

maxpiva committed:
Remove quality parameters after refactor.

majora2007 commented:
Note to self: https://github.com/Kareadita/Kavita/pull/3164 should get merged in with this Transcoding PR so any Colorscape effects can happen together.

majora2007 commented:
Just an update, I'm wrapping the current canary testing. https://github.com/Kareadita/Kavita/pull/3025 gets first priority to the canary testing (as it's likely to be a shorter test phase) then this is up. If I can, I will get this into this v0.8.4 release (as otherwise it's a bugfix release). 

Thanks for your patience. I think this will be a nice enhancement to Kavita and allow for much better format support in the future.

maxpiva commented:
Nice :) 
Hope the Panels guys add image/jxl in the accept header, since IOS supports it. Safari on iOS was working on all the new image types without transcoding.

majora2007 commented:
I reached out to Dani from Panels and they will add Accept headers to their requests so we don't need to always transcode.

majora2007 committed:
Bump versions by dotnet-bump-version.

majora2007 committed:
Lots of Bugfixes (#3308)

majora2007 commented:
@maxpiva So what was the consensus of the NetVips stuff? Wondering if I can slot this in the canary (since the Koreader one needs some more work).

maxpiva commented:
@majora2007  IDK. Asked the guy how to make it possible, but have no answer, is he/she in the discord channel?

Ref:
https://github.com/Kareadita/Kavita/pull/3141#issuecomment-2432815434

if we make it work at OS level. I could code also the netvips implementation, and we can choose at build time which one to use.

majora2007 commented:
> @majora2007 IDK. Asked the guy how to make it possible, but have no answer, is he/she in the discord channel?
> 
> Ref: [#3141 (comment)](https://github.com/Kareadita/Kavita/pull/3141#issuecomment-2432815434)
> 
> if we make it work at OS level. I could code also the netvips implementation, and we can choose at build time which one to use.

That is the maintainer of NetVips actually. They aren't in the discord. I'm not sure if it can be possible as well have native Linux builds as well. Let's move forward with canary testing for now.

majora2007 commented:
Okay the time is near. Once you get the merge conflicts out of the way, I'm good to deploy this to our canary branch and start extended testing with our dedicated testers.

maxpiva commented:
Sorry Reopen the PR by mistake

maxpiva committed:
Added Support for Q8 and Q16 ImageMagick
Q8 is a lot faster new DefineConstants Q8, Q16 and Q16HDRI can be passed to build if you want any variation, default is Q8.
Q16 and Q16HDRI is desirable when the ouput screens are >8 bit, and the initial format support more than 8Bits. IE: AVIF, HEIF or JPGXL.

maxpiva committed:
Merge from develop

maxpiva committed:
Merge branch 'Kareadita-develop' into RemoveSixLaborsAndNetVips

maxpiva commented:
Not the right or, is the other one. Also will commit some changes that
gives us 2x and 4x transcoding speed, testing right now, will or later today

El El mar, 10 dic. 2024 a la(s) 19:21, Joe Milazzo ***@***.***>
escribió:

> Reopened #3250 <https://github.com/Kareadita/Kavita/pull/3250>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Kareadita/Kavita/pull/3250#event-15608263017>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI4G3JQZRLXKMDLGWI7MRT2E5SOFAVCNFSM6AAAAABPK3SNUKVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJVGYYDQMRWGMYDCNY>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

majora2007 commented:
Sounds good. Once you are ready, give me a ping and I can get this deployed onto canary for our testers before I log off for the holiday season.

maxpiva committed:
Fix GetRGBAColors Q8 Bug
Fix merge bugs.
Added/Updated nugets, to remove transient security vulnarabilities from other nugets.

maxpiva committed:
Merge pull request #2 from Kareadita/develop

Merge from Develop

maxpiva committed:
FIX BOMs

maxpiva committed:
FIX BOMS

maxpiva committed:
FIX BOMs

maxpiva commented:
@majora2007 @DieselTech 

Merge from Develop.
Added ImageMagick Q8, Q16 and Q16HDRI support.

The original implementation was using Q16HDRI which is 4 times memory heavier and slower than the other implementations.
Now we are using Q8 which is the lighter one.

Q8 - 8 Bit per channel per pixel processing on bytes. Similar to NetVips chain.
Q16 - 16 Bit per channel per pixel processing on ushorts. 16 Bit color precision.
Q16HDRI - 16 Bit per channel per pixel processing on floats. 16 Bit color precision, more mantissa and precision.

Right now, the .NET Proxy Library, Magick.Net does not support Side by Side, you must choose one implementation or the other at build time, so, currently the user, cannot choose the ImageMagick Implementation to use, without us, using a custom Magick.Net that native load the correct library, depending on ours choice.

Nevertheless, there is 3 new Define Constants, one can use to tailor Kavita in the future.

you can add to dotnet build : 

/p:DefineConstants="Q8"
or
/p:DefineConstants="Q16"
or
/p:DefineConstants="Q16HDRI"

Is nothing is included Kavita will be built with Q8.
That will build Kavita with that implementation, code, interop code is already tested on all implementations.

Q16 will make sense, if the user has a screen capable of more than 8 Bits per pixel, and the source has more than 8 Bit per pixel also, IE: HEIF, AVIF, JPEGXL.  Some Dithering bastards, could chime in, that is also good for 8-bit sources on 10-bit screens. ;)

In Regard NetVips, Supporting the new formats requires new tooling, docker needs to install netvips native, and I was unable to make it work on normal windows, since the native nuget dosn't support the extended formats, we need to make it download it and installing manually without breaking in the Windows Build stage, or do a .bat/.powershell.

As you see I abstracted all the image processes into interfaces, so swapping engines (ImageMagick x NetVips) is possible, since Kavita is now totally abstracted of the Image Processing libraries we use, and NetVips implementation should be easy to do, since we need to refactor only the existing NetVips actions into the interfaces.

The problem is the tooling to make NetVips viable again.

So far so good on my server.

majora2007 commented:
I'm sorry @maxpiva, I created canary off the latest develop this morning and I guess there are some conflicts. Once you get them done, just dm me so I can merge and start the activities for the dedicated testers.

maxpiva commented:
Let me merge from Canary into this branch. Previously I merged and resolved from develop.

maxpiva committed:
Merged with Canaryu

maxpiva committed:
Merge branch 'Kareadita-canary' into RemoveSixLaborsAndNetVips

maxpiva commented:
@majora2007 
Merged with Canary. Runned backend tests and benchmarks. Seems fine.

FYI: Canary has differences in the frontend against develop (I just taked canary), also some version differences on Canary Kavita.Commons.csproj (not important).
Currently with Canary Fronted wont build. DO you want me to rollback frontend changes to develop?

`
Start 'Building UI'
Removing old wwwroot
Installing web dependencies

up to date, audited 725 packages in 2s

114 packages are looking for funding
  run `npm fund` for details

5 vulnerabilities (1 low, 1 moderate, 3 high)

To address all issues, run:
  npm audit fix

Run `npm audit` for details.
Building UI

> kavita-webui@0.7.12.1 prod
> npm run cache-locale-prime && ng build --configuration production && npm run minify-langs && npm run cache-locale


> kavita-webui@0.7.12.1 cache-locale-prime
> node hash-localization-prime.js

Application bundle generation failed. [1.663 seconds]

X [ERROR] Angular compilation initialization failed. [plugin angular-compiler]

  Error: The Angular Compiler requires TypeScript >=5.4.0 and <5.6.0 but 5.7.2 was found instead.
      at checkVersion (file:///C:/Sources/Repos/Kavita/UI/Web/node_modules/@angular/compiler-cli/bundles/chunk-Y3H6JOVH.js:1161:11)
      at verifySupportedTypeScriptVersion (file:///C:/Sources/Repos/Kavita/UI/Web/node_modules/@angular/compiler-cli/bundles/chunk-Y3H6JOVH.js:1165:3)
      at new NgtscProgram (file:///C:/Sources/Repos/Kavita/UI/Web/node_modules/@angular/compiler-cli/bundles/chunk-Y3H6JOVH.js:4347:7)
      at C:\Sources\Repos\Kavita\UI\Web\node_modules\@angular\build\src\tools\angular\compilation\aot-compilation.js:53:88
      at profileSync (C:\Sources\Repos\Kavita\UI\Web\node_modules\@angular\build\src\tools\esbuild\profiling.js:68:16)
      at AotCompilation.initialize (C:\Sources\Repos\Kavita\UI\Web\node_modules\@angular\build\src\tools\angular\compilation\aot-compilation.js:53:60)
      at async initialize (C:\Sources\Repos\Kavita\UI\Web\node_modules\@angular\build\src\tools\angular\compilation\parallel-worker.js:36:50)
      at async C:\Sources\Repos\Kavita\UI\Web\node_modules\@angular\build\node_modules\piscina\dist\worker.js:146:26
`

majora2007 commented:
Okay bare with me. I have no idea how canary got out of sync with develop as I deleted it this morning then recreated it off develop. 

I recreated it off of develop. It looks like there might be a few changes that need to be fixed up from your PR (or cherrypick from before the "canary" merge and remerge). Sorry about all the trouble.

xrishox commented:
wow this sounds great, i can't wait to test this out! when transcoding from jxl to a supported format, what format does it choose. at first blush it would seem like there would be no reason to not just convert to png since it's lossless, but on second thought that could be really problematic if people are on mobile and bandwidth restricted. on the other hand though encoding a lossy format to another lossy format could start to introduce artifacts. i wonder how you guys chose to address this?
